File: server/EvalutionAlghoritm/Mutation.js
import GetIdAudienceForClass from "./GetIdAudienceForClass.js";
import GetPairTypeForClass from "./GetPairTypeForClass.js";
import GetRndInteger from "./GetRndInteger.js";
import CheckPutClassForAudience from "./CheckPutClassForAudience.js";
import CheckPutClassForGroup from "./CheckPutClassForGroup.js";
import CheckPutClassForTeacher from "./CheckPutClassForTeacher.js";

export default function Mutation(schedule,
  p_genes,
  max_day,
  max_pair,
  audiences,
  mapGroupAndAG,
  mapTeacherAndAG
) {
  let new_schedule = schedule.map(sc => Object.assign({}, sc));;
  schedule.map((sch) => {
    if (Math.random() < p_genes) {
      // Получить занятия для расписания
      let clas = sch.clas;
      let pair_types = GetPairTypeForClass(clas);
      //  удалить занятие у всех групп
      let ids_ag = clas.assigned_groups.map(ag => { return ag.id });
      new_schedule = new_schedule.filter(sc => {
        if (ids_ag.find(id => id === sc.id_assigned_group))
          return false;
        else
          return true;
      })
      for (let i = 0; i < pair_types.length; i++) {
        let pair_type = pair_types[i];
        let checkAud = false;
        let checkTeach = false;
        let checkGroup = false;
        let day_week = null;
        let number_pair = null;
        let id_audience = null;
        let index = 0;

        // Пока все условия не сойдутся, рандомно выбирать параметры
        while (!checkAud || !checkTeach || !checkGroup) {
          index++;
          if (index > 1000)
            break;
          // Если есть рек время, то не выбирать случайно
          if (clas.recommended_schedules[i]) {
            day_week = clas.recommended_schedules[i].day_week;
            number_pair = clas.recommended_schedules[i].number_pair;
          }
          else {
            day_week = GetRndInteger(1, max_day);
            number_pair = GetRndInteger(1, max_pair);
          }
          id_audience = GetIdAudienceForClass(clas, audiences);
          checkAud = CheckPutClassForAudience(id_audience, new_schedule, day_week, number_pair, pair_type);
          checkTeach = CheckPutClassForTeacher(clas, new_schedule, day_week, number_pair, pair_type, mapTeacherAndAG);
          checkGroup = CheckPutClassForGroup(clas, new_schedule, day_week, number_pair, pair_type, mapGroupAndAG);
        }
        // вставить в расписание для всех групп
        clas.assigned_groups.map(ag => {
          new_schedule.unshift({
            number_pair,
            day_week,
            pair_type,
            id_assigned_group: ag.id,
            id_audience,
            clas,
          });
        })

      }
    }
  });
  return new_schedule;
}

File: server/Algorithms/IslandModelEvalutionAlgorithm/IslandModelEvalutionAlgorithm/Mutation.cpp
#include "EvalutionAlgorithm.h"

void EvalutionAlgorithm::Mutation(const int &index)
{
    // Случайное изменение пары для занятия
    int index_class = GetRndInteger(0,classes.size()-1);
    int index_pair = GetRndInteger(0,classes[index_class].schedules[index].size()-1);
    // если есть рекомендуемоемое время, то пару не менять
    int num_rec = classes[index_class].recommended_schedules.size();
    if(index_pair >= num_rec){
        int day_week = GetRndInteger(1, max_day);
        int number_pair = GetRndInteger(1,max_pair);
        int new_id_audience = GetIdAudienceForClass(classes[index_class]);
        int old_id_audience = classes[index_class].schedules[index][index_pair].id_audience;
        int pair_type = classes[index_class].schedules[index][index_pair].pair_type;
        // С шансом 0.5 менять числитель на знаменатель
        if(pair_type < 3 && GetRndDouble() <= 0.5){
            pair_type = 3 - pair_type;
        }
        classes[index_class].schedules[index][index_pair].day_week = day_week;
        classes[index_class].schedules[index][index_pair].number_pair = number_pair;
        classes[index_class].schedules[index][index_pair].pair_type = pair_type;
        // Замена ссылки для аудитории если аудитория поменялась
        if (old_id_audience != new_id_audience){
            classes[index_class].schedules[index][index_pair].id_audience = new_id_audience;
            auto ref = &classes[index_class].schedules[index][index_pair];
            auto it = find(populations[index].scheduleForAudiences[old_id_audience].begin(), populations[index].scheduleForAudiences[old_id_audience].end(), ref);
            populations[index].scheduleForAudiences[old_id_audience].erase(it);
            populations[index].scheduleForAudiences[new_id_audience].emplace(populations[index].scheduleForAudiences[new_id_audience].begin(), ref);

        }
    }
}
